### DISCUSSION TITLE
HMR in Module Federation

### DISCUSSION THREAD

**Discussion by foxaltus** [web-infra-dev/rsbuild#2802] [Q&A] [answered]
In the `module-federation` example in the repo, HMR does not seem to work when modifying the remote. Is it a limitation or a bug?

Steps:

1. Run `examples/module-federation/host` (`pnpm dev`)
2. Run `examples/modfule-federation/remote` (`pnpm dev`)
3. Open the host at http://localhost:3000/
4. Modify `Button.js` at `examples/module-federation/remote`
5. Save

=> The button is not updated on the page

**Comment by chenjiahan** [comment]
Hi, you can refer to this issue: https://github.com/module-federation/core/issues/2459

**Reply by foxaltus** [reply]
This might be my understanding of English, but I have no idea what @ScriptedAlchemy means by:

> if remote uses hosts own remote then hmr will not work. There can only be one runtime instance per build.

Should HMR work within the remotes when remotes are used in the host?

**Reply by chenjiahan** [reply]
If remotes are used in the host, HMR will not work for remotes by default.

I think this has been supported in MF 2.0, see: https://github.com/module-federation/core/discussions/2210

**Reply by foxaltus** [reply]
When you say "HMR will not work for remotes **by default**", does it imply that there is a non default way to make it work? I'm curious to see how you'd modify the official example at https://github.com/web-infra-dev/rsbuild/tree/main/examples/module-federation to make it work.

Right now, my skills are the limiting factor. All I can say is that page refresh works pretty much out of the box in our existing webpack projects with MF 1.

**Reply by foxaltus** [reply]
I see that refresh is working in the [module-federation-enhanced](https://github.com/rspack-contrib/rspack-examples/tree/main/rsbuild/module-federation-enhanced) example. The difference is that the rsbuild config uses the `ModuleFederationPlugin` from `@module-federation/enhanced/rspack` (see [docs](https://module-federation.io/guide/basic/rspack.html)), which I understand implements Module Federation 2.0. Also note that the working example is still on rsbuild 0.7.2.

So I'm going to tentatively answer that `rsbuild` does not support hot refresh in Module Federation < 2, and that it requires  `ModuleFederationPlugin` from `@module-federation/enhanced/rspack` for that.

Unfortunately this means that people who use Module Federation with webpack wont have a simple upgrade path (ideally, we would have been able to gradually upgrade MFEs to `rsbuild` while maintaining features and compatibility with the existing).

**Reply by foxaltus** [reply]
Im sorry but I need to challenge this a bit @chenjiahan. I see that the [module-federation-interop](https://github.com/rspack-contrib/rspack-examples/tree/main/rspack/module-federation-interop) for rspack has HMR working for Module Federation < 2 along with webpack, which is great for anyone looking to gradually migrate their MFEs.

What's stopping rsbuild to do the same? If nothing, please explain to modify the config in the rsbuild example. Else, why?

**Reply by chenjiahan** [reply]
I will resolve it

**Reply by chenjiahan** [reply]
Should be fixed by https://github.com/web-infra-dev/rsbuild/pull/2853 and https://github.com/web-infra-dev/rsbuild/pull/2852

**Reply by chenjiahan** [reply, marked as answer]
Hi, I just released https://github.com/web-infra-dev/rsbuild/releases/tag/v1.0.0-alpha.7 to fix the Module Federation HMR.

Now Rsbuild will set `dev.client.port` and `tools.rspack.output.uniqueName` to ensure HMR can work with remote modules.

**Reply by foxaltus** [reply]
Thank you for your support, HMR now works perfectly in the `module-federation` example 👍
I want to make more tests to check interop with webpack Module Federation, and will let you know if any issues.

**Reply by chenjiahan** [reply]
Thanks for the feedback ❤️

**Reply by foxaltus** [reply]
Hey @chenjiahan I confirm that HMR also works when the host is Rsbuild and the remote is webpack, and vice and versa. Now looks like we can have a gradual upgrade path to Rsbuild. Great work!

**Comment by sanjaykumar66** [comment]
HMR isn’t working in my setup. The host app uses the rolldown-vite bundler and loads remote apps through the @module-federation/runtime plugin. The remote app is built using Rsbuild, but HMR isn’t reflecting updates.

I can see the hot-update.js request in the network tab, but the latest code changes aren’t being applied or rendered in the UI.

I tried webpack bundler for remoteApp, in that case, HMR is working fine.
